umap/static/umap/js/modules/autocomplete.js

```diff
@@ -0,0 +1,317 @@
+      parent: document.body,
+      className: 'umap-autocomplete',
+    })
+  }
+
+  resizeContainer() {
+    const l = this.getLeft(this.input)
+    const t = this.getTop(this.input) + this.input.offsetHeight
+    this.container.style.left = `${l}px`
+    this.container.style.top = `${t}px`
+    const width = this.options.width ? this.options.width : this.input.offsetWidth - 2
+    this.container.style.width = `${width}px`
+  }
+
+  onKeyDown(e) {
+    switch (e.key) {
+      case 'Tab':
+        if (this.current !== null) this.setChoice()
+        DomEvent.stop(e)
+        break
+      case 'Enter':
+        DomEvent.stop(e)
+        this.setChoice()
+        break
+      case 'Escape':
+        DomEvent.stop(e)
+        this.hide()
+        break
+      case 'ArrowDown':
+        if (this.results.length > 0) {
+          if (this.current !== null && this.current < this.results.length - 1) {
+            // what if one result?
+            this.current++
+            this.highlight()
+          } else if (this.current === null) {
+            this.current = 0
+            this.highlight()
+          }
+        }
+        break
+      case 'ArrowUp':
+        if (this.current !== null) {
+          DomEvent.stop(e)
+        }
+        if (this.results.length > 0) {
+          if (this.current > 0) {
+            this.current--
+            this.highlight()
+          } else if (this.current === 0) {
+            this.current = null
+            this.highlight()
+          }
+        }
+        break
+    }
+  }
+
+  onKeyUp(e) {
+    const special = [
+      'Tab',
+      'Enter',
+      'ArrowLeft',
+      'ArrowRight',
+      'ArrowDown',
+      'ArrowUp',
+      'Meta',
+      'Shift',
+      'Alt',
+      'Control',
+    ]
+    if (!special.includes(e.key)) {
+      this.search()
+    }
+  }
+
+  onBlur() {
+    setTimeout(() => this.hide(), 100)
+  }
+
+  clear() {
+    this.results = []
+    this.current = null
+    this.cache = ''
+    this.container.innerHTML = ''
+  }
+
+  hide() {
+    this.clear()
+    this.container.style.display = 'none'
+    this.input.value = ''
+  }
+
+  setChoice(choice) {
+    console.log(choice)
+    choice = choice || this.results[this.current]
+    if (choice) {
+      this.input.value = choice.item.label
+      this.options.on_select(choice)
+      this.displaySelected(choice)
+      this.hide()
+      if (this.options.callback) {
+        this.options.callback.bind(this)(choice)
+      }
+    }
+  }
+
+  createResult(item) {
+    const el = DomUtil.element({
+      tagName: 'li',
+      parent: this.container,
+      textContent: item.label,
+    })
+    const result = {
+      item: item,
+      el: el,
+    }
+    DomEvent.on(el, 'mouseover', () => {
+      this.current = result
+      this.highlight()
+    })
+    DomEvent.on(el, 'mousedown', () => this.setChoice())
+    return result
+  }
+
+  resultToIndex(result) {
+    let out = null
+    this.results.forEach((item, index) => {
```

Comment:
Using `this.results.filter()` is probably more appropriate to retrieve an element from the list?

Reply:
`filters` will return an array, but `findIndex` seems to do exactly what I need, thanks for the hint!